[12.x] Add Http::batch

WendellAdriel opened the pull request:
## Overview

Inspired by the hooks from the `Bus::batch`, this PR adds a new `Http::batch` with the same hooks that are available in the `Bus::batch`.


## Use Case

Imagine that your application needs to connect to 3 different 3rd-party services to setup configurations for your customers, and you want to track the progress of the overall and also take actions after each one of these calls finish (handling success and error cases) and after everything finished you want to send a notification to the client.

With the current `Http::pool`, this is not possible. But with the new `Http::batch`, you can do that easily by setting up hooks that can run to cover all of these needs.

## Example
```php
$responses = Http::batch(fn (Batch $batch) => [
    $batch->get('http://localhost/first'),
    $batch->get('http://localhost/second'),
    $batch->get('http://localhost/third'),
])->before(function (Batch $batch) {
    // This runs before the first HTTP request is executed.
})->progress(function (Batch $batch, int|string $key, Response $response) {
    // This runs after each successful HTTP request from the Batch.
})->catch(function (Batch $batch, int|string $key, Response|RequestException $response) {
    // This runs after each failed HTTP request from the Batch.
})->then(function (Batch $batch, array $results) {
    // This runs ONLY IF all the HTTP requests from the Batch are successful and the batch is not cancelled.
})->finally(function (Batch $batch, array $results) {
    // This runs after all the HTTP requests from the Batch finish and the batch is not cancelled.
})->send();
```

## Exception

Because of the nature on how the async requests are handled, it's not possible to add requests after the batch started processing, if you attempt to do that, a `\Illuminate\Http\Client\BatchInProgressException` will be thrown.

## Helper Methods/Properties

Besides the Hooks, the `\Illuminate\Http\Client\Batch` class also provides some helper methods and properties to check details from the batch of requests:

`$batch->totalRequests` gives you the count of requests in the batch.

`$batch->pendingRequests` gives you the count of pending requests in the batch.

`$batch->failedRequests` gives you the count of failed requests in the batch.

`$batch->createdAt` gives you the datetime that the batch was created.

`$batch->cancelledAt` gives you the datetime that the batch was cancelled.

`$batch->finishedAt` gives you the datetime that the batch finished to execute all the HTTP requests.

`$batch->processedRequests()` gives you the total number of requests that have been processed by the batch.

`$batch->completion()` gives you the percentage of requests that have been processed (between 0-100).

`$batch->hasFailures()` returns true if the batch has at least one failed request.

`$batch->finished()` returns true if the batch has finished executing.

WendellAdriel committed:
Create HTTP Pool progress hook

WendellAdriel committed:
Create HTTP Pool catch hook

WendellAdriel committed:
Create HTTP Pool then hook

WendellAdriel committed:
Create HTTP Pool finally hook

hafezdivandari commented:
I wonder if it would be possible/better to have a syntax similar to `Bus::batch`, such as:

```php
Http::pool(fn (Pool $pool) => [
    $pool->get('http://localhost/first'),
    $pool->get('http://localhost/second'),
    $pool->get('http://localhost/third'),
])->before(function (Pool $pool) {
    // Logic here...
})->progress(function (Pool $pool, int|string $key, Response $response) {
    // Logic here...
})->then(function (Pool $pool, array $results) {
    // Logic here...
})->catch(function (Pool $pool, int|string $key, Response|RequestException $response) {
    // Logic here...
})->finally(function (Pool $pool, array $results) {
    // Logic here...
});
```

WendellAdriel commented:
@hafezdivandari I like the names.
I'll just wait for feedback (to see if this idea is something they would accept for the framework) before I change the implementation.

WendellAdriel committed:
Update implementation to add Http::batch

WendellAdriel committed:
Add cancel to batch implementation

WendellAdriel committed:
Refactor

WendellAdriel commented:
Hey there @taylorotwell 
I marked this PR to be reviewed again.

I updated the implementation and also updated the PR description to reflect the changes in the implementation and the use case that was faced, which I think this implementation allows an elegant and simple way to solve it.

WendellAdriel committed:
Update then and finally hooks to not run if batch is cancelled

negoziator commented:
Pull requests like this 🫶 

Great work @WendellAdriel 👌

nexxai commented:
Can the batch be added to dynamically later? Or is it that once it's created, the batch size must stay the same and a new batch may be created?

WendellAdriel commented:
> Can the batch be added to dynamically later? Or is it that once it's created, the batch size must stay the same and a new batch may be created?

Since you have access to the Batch in each of the hooks, you can add more calls by calling like:

```php
$batch->get(...);
```

taylorotwell committed:
formatting

taylorotwell committed:
formatting

timacdonald commented:
Does progress work as expected? I just tried this and found weird results.

```
Http::batch(fn ($batch) => [
    $batch->get('https://httpbin.org/delay/1'),
    $batch->get('https://httpbin.org/delay/10'),
])->before(function () {
    echo 'Before: '.date('Y-m-d H:i:s').PHP_EOL;
})->progress(function () {
    echo 'Progress: '.date('Y-m-d H:i:s').PHP_EOL;
})->send();
```

I'd expect the output the be something like:

```
Before: 2000-01-01 00:00:00
Progress: 2000-01-01 00:00:01 # 1 second later
Progress: 2000-01-01 00:00:10 # 10 seconds later
```

But instead I consistently see something like:

```
Before: 2000-01-01 00:00:00
Progress: 2000-01-01 00:00:10  # 10 seconds later
Progress: 2000-01-01 00:00:10  # 10 seconds later
```

I believe there is a problem in the foreach loop where the promises have already been combined into a single promise. Maybe I'm missing something, though.

timacdonald commented:
And even if we were to fix that, I am wondering is the `progress` method is going to do what we really want.

Imagine I have the following, where I run a slow and a fast request. 

```
Http::batch(fn ($batch) => [
    $batch->get('https://httpbin.org/delay/10'),
    $batch->get('https://httpbin.org/delay/1'),
])->before(function () {
    echo 'Before: '.date('Y-m-d H:i:s').PHP_EOL;
})->progress(function () {
    echo 'Progress: '.date('Y-m-d H:i:s').PHP_EOL;
})->send();
```

Due to PHP, I assume we would be resolving these sync, so the progress closure for the fast request would not trigger until the long request has finished?

I understand that curl multi is likely processing both of the requests under the hood at the same time, but when it comes to iterating over them in PHP, what output would we expect see?

I assume we would see:

```
Before: 2000-01-01 00:00:00
Progress: 2000-01-01 00:00:10  # 10 seconds after before
Progress: 2000-01-01 00:00:10  # 10 seconds after before
```

But I wonder if a user of the method would expect:

```
Before: 2000-01-01 00:00:00
Progress: 2000-01-01 00:00:10  # 1 second after before
Progress: 2000-01-01 00:00:10  # 10 seconds after before
```

WendellAdriel commented:
Oh, that's a good catch @timacdonald 
Thanks for that, I'll take a look into it.
I'm going to move this PR to draft while I take a look and I'll ping you when I have something if that's ok.
cc: @taylorotwell

WendellAdriel committed:
Small fix on handling result

WendellAdriel committed:
Fix issue with progress callback

WendellAdriel committed:
Fix issue with progress callback order

WendellAdriel committed:
Merge branch 'feat/http-pool-hooks' of github.com:WendellAdriel/laravel-framework into feat/http-pool-hooks

WendellAdriel committed:
Removed not used import

WendellAdriel committed:
Fix Static Analysis and skip cancel test

WendellAdriel commented:
@timacdonald @taylorotwell 
I think that I have found a way of making it work well with the progress hook.
IDK if I overcomplicated it a little tho 😅 

But with this approach, we have two things that I want to check if it's ok and we can move on like this or if we should dig more into.

1 - For making the progress hook work correctly, the way it handles the promises, there's no way to cancel the batch because it will resolve things in a non-standard order + it will dispatch all promises at the same time. Is it ok if I remove the cancel feature?

2 - Also because of this, adding new calls to the batch won't work as expected as far as I could check.

I didn't find a way to make these two things work with this new approach.
Let me know your thoughts on how I should proceed.

WendellAdriel commented:
@taylorotwell  I'll remove the cancel then, and if you feel this would be ok to merge, that's great!
Thanks again for the help and feedback on this.

WendellAdriel committed:
Remove cancel feature from Http::batch

WendellAdriel committed:
Throw exception when adding request to in progress batch

shaedrich commented:
> > Not sure about this one but since Laravel 12 requires PHP 8.2 wouldn't it be good if the public properties would be read-only? And have type declarations as close as possible?
> 
> That's a good question I had when creating the code. I saw most places were not using property types and I didn't add them, however I do agree that this could be used. I'd be happy to add typings if it's ok.

There's a first time for everything—someone has to make the first step though 😉

WendellAdriel committed:
Code Review comments

shaedrich commented:
You're welcome, thanks for applying, @WendellAdriel

taylorotwell committed:
formatting

taylorotwell committed:
remove inconsistent method

taylorotwell commented:
Thanks 👍

bretto36 commented on src/Illuminate/Http/Client/Batch.php:
```diff
@@ -0,0 +1,389 @@
+
+                    return $result;
+                },
+                'rejected' => function ($reason, $key) use ($catchCallback) {
```
@WendellAdriel There is an alert on this line with the use ($catchCallback). Is this used? Looks like it's only using $this->catchCallback and the use statement could be deleted
